splitter.go

```diff
@@ -0,0 +1,135 @@
+package plc
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// SplitReader splits reads of structs and arrays into separate reads of their components.
+type SplitReader struct {
+	Reader
+}
+
+var _ = Reader(SplitReader{}) // Compiler makes sure this type is a Reader
+
+// NewSplitReader returns a SplitReader.
+func NewSplitReader(rd Reader) SplitReader {
+	return SplitReader{rd}
+}
+
+func (r SplitReader) ReadTag(name string, value interface{}) error {
+	v := reflect.ValueOf(value)
+	if v.Kind() != reflect.Ptr {
+		return fmt.Errorf("ReadTag expects a pointer type but got %v", v.Kind())
+	}
+
+	err := error(nil)
+	switch v.Elem().Kind() {
+	case reflect.Struct:
+		str := v.Elem()
+		for i := 0; i < str.NumField(); i++ {
+			if str.Type().Field(i).PkgPath != "" {
+				continue // Type is not exported, so skip it
+			}
+
+			// Generate the name of the struct's field and recurse
+			fieldName, _ := getNameOfField(str, i)
+			if fieldName == "" {
+				continue // Can't touch that
+			}
+			fieldName = name + "." + fieldName // add prefix
+			if !str.Field(i).CanAddr() {
+				err = fmt.Errorf("Cannot address %s", fieldName)
+				break
+			}
+			fieldPointer := str.Field(i).Addr().Interface()
+
+			err = r.ReadTag(fieldName, fieldPointer)
+			if err != nil {
+				break
+			}
+		}
+	default:
+		// Just try with the underlying type
+		err = r.Reader.ReadTag(name, value)
+	}
+
+	return err
+}
+
+// SplitWriter splits writes of structs and arrays into separate writes of their components.
+type SplitWriter struct {
+	Writer
+}
+
+var _ = Writer(SplitWriter{}) // Compiler makes sure this type is a Writer
+
+// NewSplitWriter returns a SplitWriter.
+func NewSplitWriter(wr Writer) SplitWriter {
+	return SplitWriter{wr}
+}
+
+func (sw SplitWriter) WriteTag(name string, value interface{}) error {
+	v := reflect.ValueOf(value)
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem() // Naturally use what the pointer is pointing to (but only do so once)
+	}
+
+	err := error(nil)
+	switch v.Kind() {
+	case reflect.Struct:
+		str := v
+		for i := 0; i < str.NumField(); i++ {
+			if str.Type().Field(i).PkgPath != "" {
+				continue // Type is not exported, so skip it
+			}
+
+			// Generate the name of the struct's field and recurse
+			fieldName, omitempty := getNameOfField(str, i)
+			if fieldName == "" {
+				continue // Can't touch that
+			}
+			if omitempty && str.Field(i).IsZero() {
+				continue // Skip zero values
+			}
+			fieldName = name + "." + fieldName // add prefix
+			fieldPointer := str.Field(i).Interface()
+
+			err = sw.WriteTag(fieldName, fieldPointer)
+			if err != nil {
+				break
+			}
+		}
+	default:
+		// Just try with the underlying type
+		err = sw.Writer.WriteTag(name, v.Interface())
+	}
+
+	return err
+}
+
+func getNameOfField(str reflect.Value, i int) (string, bool) {
```

Comment:
Suggestion: With the caveat that I may have made a mistake here:  

The second return value right now is coded as "was omitEmpty specified on this value", which on the read path is ignored and the write path is used to skip over the value.  But, we also have a special return value (on line 123) for "" indicating we should skip over this value in both cases.  I wonder if the interface would be cleaner if the bool means "skip over this value" rather than reading either the bool value or a sentinel value in the string return value.  Then, we move the `str.Field(i).IsZero()` check into the this function. and line 123 becomes "return "", field.isZero()` 

Then, on the read and write path path, we have the same usage:

```
fieldName, ok := getNameOfField(str, i)
  if !ok {
    continue // Can't touch that
  }
```

Reply:
That could work, except it would result in effectively adding an `if omitempty && str.Field(i).IsZero() {skip}` to the read path, correct?
And that results in buggy behaviour, because if the struct you pass into ReadTag has zero elements (which it probably will, since you're likely reading into an empty struct), then those elements won't be read.
But I could pass a `shouldOmitEmpty` argument into `getNameOfField` so the read and write paths would be parallel except one would call `getNameOfField(str, i, true)` and the other would use `false`. I think that might still be cleaner than what I originally wrote?

Reply:
I pushed a commit for this. Let me know what you think.

Reply:
LGTM.